src/grid/ntv2/mod.rs

```diff
@@ -0,0 +1,97 @@
+/// - [NTv2 Archived Spec](https://web.archive.org/web/20140127204822if_/http://www.mgs.gov.on.ca:80/stdprodconsume/groups/content/@mgs/@iandit/documents/resourcelist/stel02_047447.pdf)
+/// - [ESRI](https://github.com/Esri/ntv2-file-routines/blob/master/README.md)
+/// to work with Rust Geodesy
+#[derive(Debug, Default, Clone)]
+pub struct Ntv2Grid {
+    subgrids: Vec<Ntv2SubGrid>,
+}
+
+impl Ntv2Grid {
+    pub fn new(buf: &[u8]) -> Result<Self, Error> {
+        let parser = NTv2Parser::new(buf.into());
+
+        let num_sub_grids = parser.get_u32(40) as usize;
+        if num_sub_grids != 11 && parser.cmp_str(8, "NUM_OREC") {
+            return Err(Error::Unsupported("Wrong header".to_string()));
+        }
+
+        if num_sub_grids != 1 {
+            // TODO: Add support for subgrids
+            return Err(Error::Unsupported(
+                "Contains more than one subgrid".to_string(),
+            ));
+        }
+
+        if !parser.cmp_str(56, "SECONDS") {
+            return Err(Error::Invalid("Not in seconds".to_string()));
+        }
+
+        let mut subgrids = Vec::with_capacity(num_sub_grids);
+        subgrids.push(Ntv2SubGrid::new(&parser, HEADER_SIZE)?);
+
+        Ok(Self { subgrids })
+    }
+}
+
+impl Grid for Ntv2Grid {
+    fn bands(&self) -> usize {
+        2
+    }
+
+    /// Checks if a `Coord4D` is within the grid limits +- `within` grid units
+    fn contains(&self, position: &Coor4D, within: f64) -> bool {
+        // Ntv2 spec does not allow grid extensions, so we only need to check the root grid
+        // https://web.archive.org/web/20140127204822if_/http://www.mgs.gov.on.ca:80/stdprodconsume/groups/content/@mgs/@iandit/documents/resourcelist/stel02_047447.pdf (pg 27)
+        self.subgrids[0].contains(position, within)
+    }
+
+    /// Implementation adapted from [projrs](https://github.com/3liz/proj4rs/blob/8b5eb762c6be65eed0ca0baea33f8c70d1cd56cb/src/nadgrids/grid.rs#L206C1-L252C6) && [proj4js](https://github.com/proj4js/proj4js/blob/d9faf9f93ebeccac4b79fa80f3e9ad8a7032828b/lib/datum_transform.js#L167)
+    fn interpolation(&self, coord: &Coor4D, within: f64) -> Option<Coor4D> {
+        // NOTE: This may be naive as the spec suggests the order of subgrids is not guaranteed
+        // It's ok for now because we're only supporting single subgrid grids
+        for subgrid in self.subgrids.iter().rev() {
+            if let Some(result) = subgrid.interpolation(coord, within) {
+                return Some(result);
+            }
+        }
+
+        // If we get here the grid does not contain the coordinate
+        None
+    }
+}
+
+// ----- T E S T S ---------------------------------------------------------------------
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn ntv2_grid() -> Result<(), Error> {
+        // 100800401.gsb is used in the ED50 to ETRS89 (14) transformation for the catalonia region - transformation is EPSG:5661
```

Comment:
See my comment on the gsb file itself - I will handle the attribution and copyright stuff before merging this.

The relevant info is on https://github.com/OSGeo/proj-datumgrid/issues/87 and https://www.icgc.cat/en/The-ICGC/Public-Information/Transparency/Re-use-of-the-information

Also, it is probably more proper to just say Catalonia, rather than "the catalonia region"

Reply:
Thank you, I'll move that. 

Do you have any issue with me adding the `HS2TN15_NTv2.gsb` file? It's the one I've used the most and will be able to add some more specific tests for interpolation with it.

Reply:
I cannot find HS2TN15_NTv2.gsb anywhere - do you have a link?

We would still need the URL to a clear indication (from OS, I suppose?) that the file is freely redistributable, preferably under a CC licence, such as CC-BY or CC0, with clear indications of how to properly attribute

Reply:
Ah ok, not to worry then. That .gsb is 16mb so too big. I'll find another way to test it. 

For full disclosure, I'm wanting to test the difference between my `Ntv2SubGrid` interpolation and the `BaseGrid` interpolation. When I first wrote the Ntv2 version it was because I couldn't get the `BaseGrid` interpolation result to match that of PROJ, tested via the grid in question here - you might remember [this](https://github.com/Rennzie/geodesy-wasm/pull/7#discussion_r1294267217) conversation we had over ad geodesy-wasm. My final take out from that was that the interpolations differed enough to warrent its own version in ntv2. I suspect that this difference comes down to how the grid cell is selected, but I can't confirm it 100%. At their core they both do a bilinear interpolation (I think, my math isn't strong enough) but the ntv2 version has to handle a number of nt2v specific idiosyncrasies prior to that. 

It would be nice to have one `BaseGrid` as you've suggested, I'm just not sure how we get there yet.

Reply:
@Rennzie, I'm looking into this, by comparing PROJ, ESRI and your interpolation. First by doing a very dumb interpolation to the exact lower left corner of the grid.

Your interpolation replicates the exact grid value.

PROJ is on the order of a few micrometers away, and ESRI is a fraction of a nanometer away. 
# Interpolation comparison

```console

# The PROJ value for lower left corner (40N, 0E)

$ echo 0 40 0 0 | cct -d 14 proj=hgridshift grids=100800401.gsb -- | awk "{print $2, $1}
> 39.99882421663926 -0.00120312787967

# The ESRI value for the same point, computed using https://github.com/Esri/ntv2-file-routines

$ ntv2_cvt -f 100800401.gsb 40 0
> 39.99882421665721 -0.001203127834531996

# The interpolated grid value using grid.interpolation(...)

[src\grid\ntv2\mod.rs:108] corr = Coor4D(
    [
        -0.001203127834531996,
        -0.0011757833427853057,
        0.0,
        0.0,
    ],
)

# The exact grid value 
[src\grid\ntv2\mod.rs:113] lat_corr = -0.0011757833427853057
[src\grid\ntv2\mod.rs:114] lon_corr = 0.001203127834531996

# Difference between exact grid value and grid.interpolation(...)
[src\grid\ntv2\mod.rs:118] (dlat, dlon) = (
    0.0,
    0.0,
)

# Difference, in approx nanometers, between grid.interpolation(...) and PROJ
[src\grid\ntv2\mod.rs:123] (proj_dlat, proj_dlon) = (
    1077.0539278157075,
    4513.808250783402,
)

# Difference, in approx nanometers, between grid.interpolation(...) and ESRI
[src\grid\ntv2\mod.rs:128] (esri_dlat, esri_dlon) = (
    0.1553574976997929,
    0.0,
)
```

GDAL plays "off by half" tricks and misrepresents the grid extent:

```console
$ gdalinfo 100800401.gsb

> ...
> Corner Coordinates:
> Upper Left  (  -0.0416667,  43.0416667) (  0d 2'30.00"W, 43d 2'30.00"N)
> Lower Left  (  -0.0416667,  39.9583333) (  0d 2'30.00"W, 39d57'30.00"N)
> Upper Right (   3.5416667,  43.0416667) (  3d32'30.00"E, 43d 2'30.00"N)
> Lower Right (   3.5416667,  39.9583333) (  3d32'30.00"E, 39d57'30.00"N)
> Center      (   1.7500000,  41.5000000) (  1d45' 0.00"E, 41d30' 0.00"N)
> ...
```

compared to what you read from the NTv2-header:

```console

[src\grid\ntv2\mod.rs:105] h = SubGridHeader {
    num_nodes: 1591.0,
    nlat: 43.0,
    slat: 40.0,
    wlon: 0.0,
    elon: 3.4999999999999996,
    dlat: 5.0,
    dlon: 5.0,
    num_rows: 37.0,
    row_size: 43.0,
}

```
I suppose that a sufficient amount of rounding by repeated off-by-half-corrections may lead to odd things happening in the PROJ code, although I'm not at all sure

Reply:
Thank you for looking at this 🙏! It'll be a good start point for validating BaseGrid interpolation vs Ntv2SubGrid.

Reply:
I have now updated BaseGrid to support any grid orientation (left-right, up-down).

So you should be able to create NTv2 grids from `BaseGrid::plain(...)`, by keeping the original node order, then making (lat_0, lon_0) in the header correspond to the coordinate tuple of the physically first node in the file (i.e. the lower right corner, IIRC), and the (lat_1, lon_1) in the header be the coordinate tuple of the physically last grid node in the file.

So probably start by rebasing on current `main`

Reply:
@Rennzie I have now rebased on main and in the process merged/fixup'ed a number of the 20 commits, so the rebased and cleaned up branch in my version is now just 12 commits ahead of main (was 20). All commits are still attributed to you - only the "cleanup after rebase" is mine.

All tests work, but it was a heavy rebase, so I may have missed a few things. I have a few hours to spare Thursday morning, so I will try the "ntv2 from BaseGrid::plain" thing.

Since the changes are so heavy, I'm not really sure how to proceed from here, but I'm leaning towards merging **my** branch into main, if I can get the ntv2-from-plain process to work. That will leave a fairly clear base for further work, which should then happen on a new branch.

Reply:
@busstoptaktik thanks for handling that rebase. Sounded a bit knarly. 

I'll have time tomorrow to give this a full go so I'm happy to try the "ntv2 from BaseGrid::plain" thingy. 

If you do try it, you will want to flip the sign of `lon_corr` in `parse_subgrid_grid`. You testing above showed a sign difference between the BaseGrid and my the Ntv2 interpolation which should be more correctly handled by changing the `lon_corr` sign as opposed to negating the interpolated longitude result as I'd done.

Reply:
Looking at your branch, merging it does sound like the best course of action. It's feature gated so should effect anything else, then in a new branch we can do the work to use the BaseGrid::plain for subgrids. After that I'll work on subgrid support.

Reply:
I actually removed the feature gate: I think it makes much sense to have default support for NTv2, Gravsoft and GTX